app/src/main/java/me/hufman/androidautoidrive/phoneui/SpotifyAuthorizationActivity.kt

```diff
@@ -0,0 +1,240 @@
+package me.hufman.androidautoidrive.phoneui
+
+import android.app.Activity
+import android.app.NotificationManager
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.net.Uri
+import android.os.*
+import androidx.browser.customtabs.CustomTabsIntent
+import android.util.Log
+import android.widget.Toast
+import com.adamratzman.spotify.SpotifyScope
+import com.adamratzman.spotify.getSpotifyPkceCodeChallenge
+import kotlinx.coroutines.runBlocking
+import me.hufman.androidautoidrive.AppSettings
+import me.hufman.androidautoidrive.MutableAppSettingsReceiver
+import me.hufman.androidautoidrive.music.spotify.SpotifyWebApi
+import me.hufman.androidautoidrive.music.controllers.SpotifyAppController
+import me.hufman.androidautoidrive.music.spotify.SpotifyAuthStateManager
+import net.openid.appauth.*
+import net.openid.appauth.browser.BrowserWhitelist
+import net.openid.appauth.browser.VersionedBrowserMatcher
+import net.openid.appauth.connectivity.DefaultConnectionBuilder
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.random.Random
+
+/**
+ * Activity that is used to perform the Spotify authorization process and update the [AuthState] with
+ * the resulting authorization code. The activity will be closed once the process has either been
+ * cancelled, failed, or succeeds.
+ */
+class SpotifyAuthorizationActivity: Activity() {
+	companion object {
+		const val TAG = "AuthorizationActivity"
+
+		const val REQUEST_CODE_SPOTIFY_LOGIN = 2356
+
+		const val EXTRA_AUTHORIZATION_RESULT = "authorizationResult"
+		const val AUTHORIZATION_CANCELED = 1
+		const val AUTHORIZATION_SUCCESS = 0
+		const val AUTHORIZATION_FAILED = -1
+
+		val CODE_VERIFIER = generateCodeVerifierString()
+
+		/**
+		 * Generates a 45 character random alpha numeric string to be used for the code verifier in
+		 * the PKCE authentication flow.
+		 */
+		private fun generateCodeVerifierString(): String {
+			val charPool = ('a'..'z') + ('A'..'Z') + ('0'..'9')
+			return (1..45).map { charPool[Random.nextInt(0, charPool.size)] }.joinToString("")
+		}
+	}
+
+	private val authRequest = AtomicReference<AuthorizationRequest>()
+	private val authIntent = AtomicReference<CustomTabsIntent>()
+	private lateinit var authService: AuthorizationService
+	private var authIntentLatch = CountDownLatch(1)
+	private lateinit var authStateManager: SpotifyAuthStateManager
```

Comment:
Instead of `lateinit var` I prefer the pattern of `val authStateManager by lazy { }` to use the val immutable enforcement

Reply:
Oh, didn't even realize those were an option in Kotlin! Updated these lateinits to use that by lazy { }